File: Leetcode/Hard/99.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    TreeNode *first, *second, *previous;
    
    void inOrder(TreeNode* root){
        if(!root) return;
        
        inOrder(root->left);
        
        if(previous != NULL and root->val < previous->val){
            if(first == NULL){
                first = previous;
            }
            second = root;
        }
        
        previous = root;
        inOrder(root->right);
    }
    
    void recoverTree(TreeNode* root) {
        first = second = previous = NULL;
        inOrder(root);
        swap(first->val, second->val);
    }
};

File: Leetcode/Hard/99.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 var(
    first *TreeNode
    second *TreeNode
    previous *TreeNode
)


func recoverTree(root *TreeNode)  {
    first, second, previous = nil, nil, nil
    
    inOrderTraversal(root)
    
    t := first.Val
    first.Val = second.Val
    second.Val = t
}

func inOrderTraversal(root *TreeNode) {
    if root == nil {
        return
    }
    
    inOrderTraversal(root.Left)
    
    if previous != nil && root.Val < previous.Val {
        if first == nil {
            first = previous
        }
        second = root
    }
    
    previous = root
    inOrderTraversal(root.Right)
}

